Refactor and use `ScrollList` component in more places

mvdbeek opened the pull request:
# Custom Tools Panel updates
The Custom Tools panel now uses `GCard` in the existing `ScrollList`. This allows for nicer action buttons and badges.

| Before | After |
| ---- | ---- |
| <img width="337" height="864" alt="firefox_VGv46860af" src="https://github.com/user-attachments/assets/c7c952f6-f269-4b45-8683-de82a2f38c19" /> | <img width="337" height="864" alt="AiDPsKKJ1L" src="https://github.com/user-attachments/assets/147fcda5-0358-4047-9622-b2460c487083" /> |


# HistoryScrollList updates
The `HistoryScrollList` component now uses `ScrollList` and `GCard`.

## Current History Selector

| Before | After |
| ---- | ---- |
| <img width="722" height="842" alt="firefox_xlPpzFX9n2" src="https://github.com/user-attachments/assets/69838e3e-a028-4123-a57b-d372dc353c59" /> | <img width="722" height="842" alt="H56w2jcrTl" src="https://github.com/user-attachments/assets/d46f13ae-c2d0-45be-9b25-5bead6c50878" /> |


## History Multiview Selector Modal

| Before | After |
| ---- | ---- |
| <img width="722" height="841" alt="GM6SNXLDA2" src="https://github.com/user-attachments/assets/c1a652dd-f4a5-48c3-8916-a29aeffebacd" /> | <img width="722" height="841" alt="4kgGDwdUhs" src="https://github.com/user-attachments/assets/75cb5821-4923-446f-a9dd-5f3a0ea16493" /> |





## History Multiview Selector Panel

| Before | After |
| ---- | ---- |
| <img width="334" height="863" alt="firefox_PqwcJxsp3v" src="https://github.com/user-attachments/assets/d75c550a-76a2-4cca-8402-db5ce5b187cb" /> | <img width="334" height="863" alt="pp5RdlV7e2" src="https://github.com/user-attachments/assets/dec85a39-f63a-468a-ab65-d31154c2cfec" /> |



## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

ahmedhamidawan committed:
try out generic component typing in `ScrollList`, fix/finalize in vue 3

Generic prop typing https://blog.vuejs.org/posts/vue-3-3#generic-components will allow us to have generic types in components. This is a workaround for now, but it does help type the `item` slot nicely for the parents that use `ScrollList` for now.

ahmedhamidawan committed:
use `GCard` in `UserToolPanel`

ahmedhamidawan committed:
remove the eye icon from scroll lists; current/active indicators are better

ahmedhamidawan committed:
indicate scroll list loading/busy via spinner icon

ahmedhamidawan committed:
use `ScrollList` in `HistoryScrollList`

Had to modify `ScrollList` to allow two options
1. The existing implementation with a provided (required) `loader` function, that loads items onto the local state (local refs).
2. OR, now we've made the `loader` optional, and we require the items, total count and a busy state as props if the loader isn't provided.

This way, scroll list items can be loaded in the props, or in the component itself.

ahmedhamidawan committed:
use `GCard` in `HistoryScrollList`

ahmedhamidawan committed:
make all `v-b-tooltip`s noninteractive in `GCard`

ahmedhamidawan committed:
change `GCard` bottom padding to margin for better click separation

Due to padding, you would still be clicking the card above if there's another card below.

ahmedhamidawan committed:
use base link and button components in `SelectorModal`

ahmedhamidawan committed:
fix `SelectorModal` jest for the scroll list and card updates

ahmedhamidawan committed:
add a "will be unpinned" badge to history multiview selector modal

ahmedhamidawan committed:
improve styling for scroll list count indicators; use `list-end` class

ahmedhamidawan committed:
use `ScrollList` in `VisualizationPanel`

itisAliRH commented on client/src/components/History/HistoryScrollList.vue:
```diff
@@ -1,30 +1,21 @@
 import { useUserStore } from "@/stores/userStore";
-import localize from "@/utils/localization";
 
+import type { CardAction, CardBadge } from "../Common/GCard.types";
```
```suggestion
import type { CardAction, CardBadge } from "@/components/Common/GCard.types";
```

itisAliRH commented on client/src/components/History/HistoryScrollList.vue:
```diff
@@ -1,30 +1,21 @@
-import GButtonGroup from "@/components/BaseComponents/GButtonGroup.vue";
-import Heading from "@/components/Common/Heading.vue";
+import GCard from "../Common/GCard.vue";
+import ScrollList from "../ScrollList/ScrollList.vue";
```
```suggestion
import GCard from "@/components/Common/GCard.vue";
import ScrollList from "@/components/ScrollList/ScrollList.vue";
```

itisAliRH commented on client/src/components/History/HistoryScrollList.vue:
```diff
@@ -225,154 +181,151 @@ async function loadMore(noScroll = false) {
+        </template>
+
+        <template v-slot:item="{ item: history }">
+            <!-- TODO: Long history names should be truncated, esp in multiview panel -->
```
We may need this in other places as well. I think we should create it in GCard and control it using a prop.

ahmedhamidawan replied:
This was tricky 😓 

Final result is, I've had to compromise with the only apparent styling bug i.e. a badly positioned tooltip:

<img width="367" height="364" alt="firefox_C5bNjUHLMV" src="https://github.com/user-attachments/assets/a099c36b-3beb-49fe-870d-df03d841acc7" />

### There's a comment that explains why I did this:
(I guess ~height setting~ ✖️  display setting* ✔️ )

https://github.com/galaxyproject/galaxy/blob/8901374d695336fa520fada762a312a758709ee6/client/src/components/Common/GCard.vue#L717-L727

Any tips or is this good as is? Feel free to push etc too!

itisAliRH replied:
@ahmedhamidawan I could fix the tooltip issue.

itisAliRH commented on client/src/components/History/HistoryScrollList.vue:
```diff
@@ -225,154 +181,151 @@ async function loadMore(noScroll = false) {
+                        :disabled="true"
+                        :max-visible-tags="isMultiviewPanel ? 1 : 10"
+                        @tag-click="setFilterValue('tag', $event)" />
+                </template>
```
`GCard` can do this part using its props. e.g. [WorkflowCard](https://github.com/galaxyproject/galaxy/blob/dev/client/src/components/Workflow/List/WorkflowCard.vue#L128) has tags.

ahmedhamidawan replied:
Ah! You're right, my bad, will use the props for tags.

itisAliRH replied:
NP! It also works for description.

itisAliRH commented on client/src/components/Panels/UserToolPanel.vue:
```diff
@@ -10,10 +8,12 @@ import { useRoute, useRouter } from "vue-router/composables";
 
+import GButton from "../BaseComponents/GButton.vue";
+import GButtonGroup from "../BaseComponents/GButtonGroup.vue";
+import GCard from "../Common/GCard.vue";
```
```suggestion
import GButton from "@/components/BaseComponents/GButton.vue";
import GButtonGroup from "@/components/BaseComponents/GButtonGroup.vue";
import GCard from "@/components/Common/GCard.vue";
```

itisAliRH commented on client/src/components/ScrollList/ScrollList.vue:
```diff
@@ -1,131 +1,134 @@
 import { useAnimationFrameScroll } from "@/composables/sensors/animationFrameScroll";
 
+import GButton from "../BaseComponents/GButton.vue";
+import LoadingSpan from "../LoadingSpan.vue";
```
```suggestion
import GButton from "@/components/BaseComponents/GButton.vue";
import LoadingSpan from "@/components/LoadingSpan.vue";
```

itisAliRH commented on client/src/components/Visualizations/VisualizationPanel.vue:
```diff
@@ -1,15 +1,14 @@
 
 import { getTestExtensions } from "./utilities";
 
+import ScrollList from "../ScrollList/ScrollList.vue";
```
```suggestion
import ScrollList from "@/components/ScrollList/ScrollList.vue";
```

itisAliRH reviewed (changes requested):
Thank you @ahmedhamidawan for working on this. I added some comments for your review.

ahmedhamidawan committed:
Fix shortened component imports

Co-authored-by: Alireza Heidari <itisalirh@gmail.com>

ahmedhamidawan committed:
sort component imports

ahmedhamidawan committed:
use the tags prop for `GCard` instead of using tags separately

Co-authored-by: Alireza Heidari <itisalirh@gmail.com>

ahmedhamidawan committed:
make `VisualizationPanel` a div when in center, `ActivityPanel` otherwise

ahmedhamidawan committed:
add a `title-n-lines` prop to `GCard`

ahmedhamidawan committed:
truncate `InvocationScrollList` card titles too; fix inline styling

itisAliRH committed:
Refactors card title rendering

Simplifies the structure of the card title by removing redundant, wrapping elements and unnecessary CSS classes.
Adds `g-card-title-truncate` class for handling title truncation consistently for title.
Eliminates obsolete styles related to title actions
and adjusts layout for better alignment and maintainability.

itisAliRH committed:
Refactor GCard to use computed property for title line truncation

itisAliRH committed:
Simplifies history scroll list card description handling

Refactors to directly bind the description property, removing unnecessary template logic for annotations.

itisAliRH committed:
Prevents event propagation on toggle button click

Adds `.prevent.stop` modifiers to the click event handler of the toggle button to prevent default browser behavior and stop event propagation.

This ensures that clicking the button does not trigger unintended parent or sibling event listeners (e.g. item click in the GCard).
